1802 docs makeover (#1892)
software-mansion/protostar: website/docs/cairo-1/05-testing/02-integration-testing.md

---
sidebar_label: Integration testing
---

# Integration testing

> Using [unit testing](./01-unit-testing.md) as much as possible is a good practice, as it makes your test suites run faster. However, when writing smart contracts you often want your test to cover on-chain state and interactions between multiple contracts.

Reviewer: ```suggestion
Using [unit testing](./01-unit-testing.md) as much as possible is a good practice, as it makes your test suites run faster. However, when writing smart contracts you often want your test to cover the on-chain state and interactions between multiple contracts.
```

> In this section you will learn how to deploy and interact with a smart contract in Protostar for testing purposes. 

Reviewer: ```suggestion
In this section, you will learn how to deploy and interact with a smart contract in Protostar for testing purposes. 
```

## How to test a contract

> To test a contract you need to use important Protostar feature, [cheatcodes](./03-cheatcodes.md). Cheatcodes are additional library functions which Protostar exposes to help you with testing.

Reviewer: ```suggestion
To test a contract you need to use an important Protostar feature: [cheatcodes](./03-cheatcodes.md). They are additional library functions that Protostar exposes to help you with testing.
```

> Let's write a test which deploys and calls a contract. First let's define our contract 

Reviewer: ```suggestion
Let's write a test that deploys and calls a contract. First, let's define our contract 
```

```cairo title="Deployed contract"
#[contract]
mod MinimalContract {
    #[external]
    fn panic_with(panic_data: Array::<felt252>) {
        panic(panic_data);
    }
}
```

> We can write a test which deploys and calls this contract. Let's create a file `test_contract.cairo`:

Reviewer: ```suggestion
We can write a test that deploys and calls this contract. Let's create a file `test_contract.cairo`:
```

> We can write a test which deploys and calls this contract. Let's create a file `test_contract.cairo`:

Reviewer: ```suggestion
We can write a test that deploys and calls this contract. Let's create a file `test_contract.cairo`:
```

```cairo title="Example"
use array::ArrayTrait;
use result::ResultTrait;

#[test]
fn test_deploy() {
    let deployed_contract_address = deploy_contract('minimal', ArrayTrait::new()).unwrap();

>     TODO (add calling code)

Reviewer: ```suggestion
    // TODO (add calling code)
```
They may wanna check if deploying itself works so we might not want to break compilation

Author: I will add this code here in this PR

}
```

> This cheatcode will declare and deploy given contract.

Reviewer: ```suggestion
This cheatcode will declare and deploy the given contract.
```
Also prepare, no?

## Under the hood
You may ask, where the contract has been deployed? Protostar comes with a local Starknet instance which you can use to test your contracts. 
To encourage good testing practices, each test case starts with a fresh Starknet instance. 

> When Starknet instance is accessed through cheatcodes, it is analogous to accessing real Starknet through gateway. Example consequence of this behavior is that `get_caller_address` will return `0` in the called contract.

Reviewer: ```suggestion
When the Starknet instance is accessed through cheatcodes, it is analogous to accessing real Starknet through the gateway. An example consequence of this behavior is that `get_caller_address` will return `0` in the called contract.
```

> When Starknet instance is accessed through cheatcodes, it is analogous to accessing real Starknet through gateway. Example consequence of this behavior is that `get_caller_address` will return `0` in the called contract.

Author: I think analogous is better here. There is no equivalence

## Transaction reverts

[deploy](./cheatcodes-refernce/deploy.md), [invoke](./cheatcodes-refernce/invoke.md) and [call](./cheatcodes-refernce/call.md) execute code on chain which can be reverted.

> In such case, they return `RevertedTransaction` structure. You can use it, for example, to verify if your contract reverts transaction in certain scenario.

Author: https://github.com/software-mansion/protostar/issues/1899

> In such case, they return `RevertedTransaction` structure. You can use it, for example, to verify if your contract reverts transaction in certain scenario.

Reviewer 2: Isn't it documented already? Although I believe we can move it to separate section and improve it a bit.
https://docs.swmansion.com/protostar/docs/tutorials/cairo-1-support/cheatcodes/invoke#handling-invoke-failures

> In such case, they return `RevertedTransaction` structure. You can use it, for example, to verify if your contract reverts transaction in certain scenario.

Reviewer 3: I think #1850 should resolve this issue with docs, opening issue is enough by me for now.
